sql/query/expr.go

```
@@ -0,0 +1,660 @@
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal |= *right.UintVal
+		case intOp:
+			*left.IntVal |= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Bitxor:
+		typ, left, right, err = prepareBinaryArgs(intOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal ^= *right.UintVal
+		case intOp:
+			*left.IntVal ^= *right.IntVal
+		}
+		return left, nil
+
+	case parser.Plus:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal += *right.UintVal
+		case intOp:
+			*left.IntVal += *right.IntVal
+		case floatOp:
+			*left.FloatVal += *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Minus:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			// If the unsigned subtraction would result in a negative number, convert
+			// to signed values.
+			if *right.UintVal > *left.UintVal {
+				v := -int64(*right.UintVal - *left.UintVal)
+				left = sqlwire.Datum{IntVal: &v}
+			} else {
+				*left.UintVal -= *right.UintVal
+			}
+		case intOp:
+			*left.IntVal -= *right.IntVal
+		case floatOp:
+			*left.FloatVal -= *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Mult:
+		typ, left, right, err = prepareBinaryArgs(floatOp, left, right)
+		if err != nil {
+			return null, err
+		}
+		switch typ {
+		case uintOp:
+			*left.UintVal *= *right.UintVal
+		case intOp:
+			*left.IntVal *= *right.IntVal
+		case floatOp:
+			*left.FloatVal *= *right.FloatVal
+		}
+		return left, nil
+
+	case parser.Div:
+		// Division always operates on floats. TODO(pmattis): Is this correct?
+		left, err = left.ToFloat()
+		if err != nil {
+			return null, err
+		}
+		right, err = right.ToFloat()
+		if err != nil {
+			return null, err
+		}
+		*left.FloatVal /= *right.FloatVal
```

Hmm, I think it's better to return an error in such cases. Postgres return `division by 0`. Thoughts?